File: Algo/Leetcode/322CoinChange.cpp
// dp. O(nk) time, O(n) space
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if(amount==0) return 0;
        vector<int> dp(amount+1, amount+1);
        dp[0] = 0;

        for(int i=1; i<=amount; i++){
            for(auto c:coins){
                if(c<=i)
                    dp[i] = min(dp[i-c]+1, dp[i]);
            }
        }
        return dp[amount]>amount?-1:dp[amount];
    }
};

File: Algo/Leetcode/322CoinChange.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        if amount == 0:
            return 0

        dp = [amount+1] * (amount+1)        # amount + 1 means can not reach
        dp[0] = 0
        for i in range(1, amount+1):
            dp[i] = min([dp[i]] + [dp[i-c]+1 if c <= i else amount+1 for c in coins])

        return dp[-1] if dp[-1] != amount+1 else -1